File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        subset = []
        candidates.sort()

        def bt(i, sub_sum):
            if sub_sum == target:
                res.append(subset.copy())
                return
            
            if sub_sum >= target or i >= len(candidates):
                return

            subset.append(candidates[i])
            bt(i + 1, sub_sum + candidates[i])
            subset.pop()

            while i + 1 < len(candidates) and candidates[i] == candidates[i+1]:
                i += 1
            bt(i + 1, sub_sum)

        bt(0, 0)
        return res

File: Data Structures & Algorithms/subsets-ii/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsetsWithDup(nums) {
        function bt(ind) {
            if (ind === nums.length) {
                res.push([...subset]);
                return;
            }

            subset.push(nums[ind]);
            bt(ind+1);
            subset.pop(nums[ind]);
            
            while (ind+1 < nums.length && nums[ind] === nums[ind+1]) {
                ind++;
            }
            bt(ind+1);
        }

        const res = [];
        const subset = [];
        nums.sort((a, b) => a-b);
        bt(0);
        return res;
    }
}
